File: data/sliding-window/find-all-anagrams.cpp
#include <vector>
#include <string>

using namespace std;

vector<int> findAnagrams(string s, string p) {
    if (p.length() > s.length()) return {};
    
    vector<int> p_count(26, 0), s_count(26, 0), res;
    for (int i = 0; i < p.length(); ++i) {
        p_count[p[i] - 'a']++;
        s_count[s[i] - 'a']++;
    }
    
    if (p_count == s_count) res.push_back(0);
    
    for (int i = p.length(); i < s.length(); ++i) {
        s_count[s[i] - 'a']++;
        s_count[s[i - p.length()] - 'a']--;
        if (p_count == s_count) res.push_back(i - p.length() + 1);
    }
    
    return res;
}

File: data/sliding-window/find-all-anagrams.py
from collections import Counter

def findAnagrams(s: str, p: str) -> list[int]:
    if len(p) > len(s): return []
    
    p_count = Counter(p)
    s_count = Counter(s[:len(p)])
    
    res = []
    if p_count == s_count:
        res.append(0)
        
    for i in range(len(p), len(s)):
        s_count[s[i]] += 1
        s_count[s[i - len(p)]] -= 1
        if s_count[s[i - len(p)]] == 0:
            del s_count[s[i - len(p)]]
            
        if p_count == s_count:
            res.append(i - len(p) + 1)
            
    return res
